File: producer.js
/*
 * Requires node-mongodb-native to be installed locally.
 * Run command: node producer.js
 */

var cluster = require('cluster'),
    MongoClient = require('mongodb').MongoClient;

var COUNT = 100000; // must match consumer
var count = 0;

MongoClient.connect('mongodb://localhost:27017/default', { "server": { "poolSize": 1024 }}, function (err, db) {
  if (err) throw err;
  db.collection('buffer', function (err, buffer) {
    if (err) throw err;

    var doInsert = function () {
      buffer.insert({ "event": "fast", "timestamp": new Date() }, function (err) {
        if (err) throw err;
        count ++;
        if (count === COUNT) {
          console.log('total:', count);
          console.log('msg/s:', count / (new Date().getTime() - start.getTime()) * 1000);
          process.exit();
        }
      });
    };

    var start = new Date();
    for (var i = 0; i <= COUNT; i++) {
      if (i % 32 === 0) {
        process.nextTick(doInsert);
      }
      else {
        doInsert();
      }
    }

    process.on('exit', function () {
      db.close();
    });

    process.on('SIGINT', function () {
      process.exit();
    });
  });
});

File: producer.cpp
/*
 * You must compile the MongoDB C++ driver locally before compiling.
 * Build command: g++ producer.cpp -pthread -lmongoclient -lboost_thread-mt -lboost_filesystem -lboost_program_options -lboost_system -L./mongo-client-install/lib/ -I./mongo-client-install/include/ -o producer
 * Run command: ./producer
 */

#include <cstdlib>
#include <iostream>
#include "mongo/client/dbclient.h"

using namespace mongo;

int main(int argc, char* argv[]) {
    DBClientConnection conn;
    conn.connect("localhost");

    const uint32_t COUNT = 100000;

    const Date_t start = jsTime();
    for(uint32_t i = 0; i < COUNT; i++) {
        conn.insert("default.buffer", BSON( "event" << "fast" << "timestamp" << jsTime() ), 0);
    }

    std::cout << "total: " << COUNT << std::endl;
    std::cout << "msg/s: " << (float)COUNT / (float)(jsTime().millis - start.millis) * 1000.0 << std::endl;

    return 0;
}
